Include/object.h

```diff
@@ -530,13 +616,28 @@ static inline void Py_INCREF(PyObject *op)
     // Stable ABI for Python 3.10 built in debug mode.
     _Py_IncRef(op);
 #else
-    _Py_INCREF_STAT_INC();
     // Non-limited C API and limited C API for Python 3.9 and older access
     // directly PyObject.ob_refcnt.
+#if SIZEOF_VOID_P > 4
+    // Portable saturated add, branching on the carry flag and set low bits
+    PY_UINT32_T new_refcnt;
+    PY_UINT32_T cur_refcnt = _Py_CAST(PY_UINT32_T, op->ob_refcnt);
+    new_refcnt = cur_refcnt + 1;
+    if (new_refcnt < cur_refcnt) {
+        return;
+    }
```

Comment:
Glad you asked! Yes, I definitely have, this is what I converged to after dozens of iterations!  We are actually already using the fact that we almost always have the high 32 bits set to zero! In fact, originally I was handwriting this in assembly but it turns out that both GCC and LLVM already handle neatly.

Theres two major tricks here. First, the explicit cast of the refcount to a 32bit number results in loading the value to a 32bit register instead of a 64bit one (i.e eax instead of rax). Knowing this, we can then do an add on the 32 bit register, which will saturate if it's already 0xFFFFFF (the immortal value).

The second major trick is the if check, by adding and then checking if it's smaller than the original value, it results in a single `jc` instruction, rather than the `test` -> `js` pattern of a normal if check. This bypasses the test instruction, resulting in one less overall instruction for Py_INCREF which is a major performance improvement. Here's how the assembly ends up looking like:

With saturated add:
```
        mov     eax, DWORD PTR [rdi]
        add     eax, 1
        jc      .L6
        mov     DWORD PTR [rdi], eax
```

Without saturated add:
```
        mov     rax, QWORD PTR [rdi]
        cmp     rax, 16777215
        je      .L14
        add     rax, 1
        mov     QWORD PTR [rdi], rax
```


Feel free to play around with various other options (I've already populated it with a couple of options): https://godbolt.org/z/4oz7YEG1a

For context, this exact structure is what has improved performance the most both in valgrind (in instruction count and cache misses) and in the pyperformance benchmark suite (overall e2e time).